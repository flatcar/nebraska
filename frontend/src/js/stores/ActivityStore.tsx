import Link from '@material-ui/core/Link';
import React from 'react';
import { Link as RouterLink } from 'react-router-dom';
import _ from 'underscore';
import API from '../api/API';
import { Activity } from '../api/apiDataTypes';
import Store from './BaseStore';

class ActivityStore extends Store {
  activity: {[key: string]: Activity[]} | null | never[];
  constructor() {
    super();
    this.activity = null;
    this.getActivity();

    setInterval(() => {
      this.getActivity();
    }, 60 * 1000);
  }

  getCachedActivity() {
    return this.activity;
  }

  getActivity() {
    API.getActivity()
      .then(activity => {
        this.activity = _.isNull(activity) ? [] : this.sortActivityByDate(activity);
        this.emitChange();
      })
      .catch((error) => {
        if (error.status === 404) {
          this.activity = [];
          this.emitChange();
        }
      });
  }

  sortActivityByDate(entries: Activity[]) {
    const sortedEntries: {
      [key: string]: Activity[];
    } = {};

    entries.forEach(entry => {
      const createdDate = new Date(entry.created_ts);
      const date = createdDate.toLocaleDateString('default', {day: 'numeric', weekday: 'short', month: 'short', year: 'numeric'});
      if (_.has(sortedEntries, date)) {
        sortedEntries[date].push(entry);
      }
      else {
        sortedEntries[date] = [entry];
      }
    });

    return sortedEntries;
  }

  getActivityEntryClass(classID: number, entry: Activity) {
    const instancePath = `/apps/${entry.app_id}/groups/${entry.group_id}/instances/${entry.instance_id}?period=1d`;

    const classType: {
      [key: string]: {
        type: string;
        appName: string;
        groupName: string | null;
        channelName: string | null;
        description: string | React.ReactElement;
      };
    } = {
      1: {
        type: 'activityPackageNotFound',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: "An update request could not be processed because the group's channel is not linked to any package"
      },
      2: {
        type: 'activityRolloutStarted',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: 'Version ' + entry.version + ' roll out started'
      },
      3: {
        type: 'activityRolloutFinished',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: 'Version ' + entry.version + ' successfully rolled out'
      },
      4: {
        type: 'activityRolloutFailed',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: 'There was an error rolling out version ' + entry.version + " as the first update attempt failed. Group's updates have been disabled"
      },
      5: {
        type: 'activityInstanceUpdateFailed',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: (
          <React.Fragment>
            Instance <Link component={RouterLink} to={instancePath}>{entry.instance_id}</Link>
            {' '} reported an error while processing update to version {entry.version}
          </React.Fragment>
        )
      },
      6: {
        type: 'activityChannelPackageUpdated',
        appName: entry.application_name,
        groupName: entry.group_name,
        channelName: entry.channel_name,
        description: 'Channel ' + entry.channel_name + ' is now pointing to version ' + entry.version
      }
    };

    const classDetails = classID ? classType[classID] : classType[1];
    return classDetails;
  }

  getActivityEntrySeverity(severityID: number) {

    const severityType: {
      [key: string]: {
        type: string;
        className: string;
        icon: string;
      };
    } = {
      1: {
        type: 'activitySuccess',
        className: 'success',
        icon: 'fa-check'
      },
      2: {
        type: 'activityInfo',
        className: 'info',
        icon: 'fa-info'
      },
      3: {
        type: 'activityWarning',
        className: 'warning',
        icon: 'fa-exclamation'
      },
      4: {
        type: 'activityError',
        className: 'error',
        icon: 'fa-close'
      }
    };

    const severityInfo = severityID ? severityType[severityID] : severityType[1];
    return severityInfo;
  }
}

export default ActivityStore;
