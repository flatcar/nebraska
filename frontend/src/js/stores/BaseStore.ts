import EventEmitter from 'events';

var CHANGE_EVENT = 'change';

class Store extends EventEmitter {

  constructor() {
    super();
    this.setMaxListeners(100);
  }

  emitChange() {
    this.emit(CHANGE_EVENT);
  }

  addChangeListener(callback: () => void) {
    this.on(CHANGE_EVENT, callback);
  }

  removeChangeListener(callback: () => void) {
    this.removeListener(CHANGE_EVENT, callback);
  }
}

export default Store;
