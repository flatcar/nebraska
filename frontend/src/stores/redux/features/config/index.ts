import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface NebraskaConfig {
  title: string;
  nebraska_config: string;
  nebraska_version: string;
  login_url: string;
  auth_mode: string;
  header_style: string;
  [prop: string]: any;
}

const initialState: NebraskaConfig = {} as NebraskaConfig;

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {
    setConfig: (state, action: PayloadAction<NebraskaConfig>) => {
      Object.assign(state, action.payload);
    },
  },
});

export const { setConfig } = configSlice.actions;
export default configSlice.reducer;
