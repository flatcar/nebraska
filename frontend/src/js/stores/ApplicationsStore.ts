import _, { Collection, List } from 'underscore';
import API from '../api/API';
import { Application, Channel, Group, Package } from '../api/apiDataTypes';
import Store from './BaseStore';
import { setUser } from './redux/features/user';
import store from './redux/store';

class ApplicationsStore extends Store {
  applications: Application[] | null;
  constructor() {
    super();
    this.applications = null;
    this.getApplications();

    setInterval(() => {
      this.getApplications();
    }, 60 * 1000);
  }

  // Applications

  getCachedApplications() {
    return this.applications;
  }

  getCachedApplication(applicationID: string) {
    const app = _.findWhere(this.applications as Collection<any>, { id: applicationID });
    if (!app) {
      return null;
    }
    return { ...app };
  }

  getApplications() {
    API.getApplications()
      .then(response => {
        this.applications = response.applications;
        this.emitChange();
      })
      .catch(error => {
        switch (error.status) {
          case 404:
            this.applications = [];
            this.emitChange();
            break;
          case 401:
            store.dispatch(setUser({ authenticated: false }));
            break;
          default:
            console.debug('Error', error);
        }
      });
  }

  getApplication(applicationID: string) {
    API.getApplication(applicationID).then(application => {
      if (this.applications) {
        const applicationItem = application;
        const index = _.findIndex(this.applications, { id: applicationID });
        if (index >= 0) {
          this.applications[index] = applicationItem;
        } else {
          this.applications.unshift(applicationItem);
        }
        this.emitChange();
      }
    });
  }

  getCachedPackages(applicationID: string) {
    const app = _.findWhere(this.applications as Collection<any>, { id: applicationID });
    const packages = app ? app.packages : [];
    return packages;
  }

  getCachedChannels(applicationID: string) {
    const app = _.findWhere(this.applications as Collection<any>, { id: applicationID });
    const channels = app ? app.channels : [];
    return channels;
  }

  createApplication(
    data: Pick<Application, 'name' | 'description' | 'product_id'>,
    clonedApplication: string
  ) {
    return API.createApplication(data, clonedApplication).then(application => {
      const applicationItem = application;
      if (this.applications) {
        this.applications.unshift(applicationItem);
        this.applications = [...this.applications];
        this.emitChange();
      }
    });
  }

  updateApplication(applicationID: string, data: any) {
    data.id = applicationID;

    return API.updateApplication(data).then(application => {
      const applicationItem = application;
      const applicationToUpdate = _.findWhere(this.applications as Collection<any>, {
        id: applicationItem.id,
      });

      ['name', 'description', 'product_id'].forEach(item => {
        applicationToUpdate[item] = applicationItem[item];
      });

      this.emitChange();
    });
  }

  getAndUpdateApplication(applicationID: string) {
    API.getApplication(applicationID).then(application => {
      const applicationItem = application;
      const index = _.findIndex(this.applications as List<any>, { id: applicationID });
      if (this.applications) {
        this.applications[index] = applicationItem;
        this.emitChange();
      }
    });
  }

  deleteApplication(applicationID: string) {
    API.deleteApplication(applicationID).then(() => {
      this.applications = _.without(
        this.applications as List<any>,
        _.findWhere(this.applications as Collection<any>, { id: applicationID })
      );
      this.emitChange();
    });
  }

  // Groups

  createGroup(data: Group) {
    return API.createGroup(data).then(group => {
      const groupItem = group;
      const applicationToUpdate = _.findWhere(this.applications as Collection<any>, {
        id: groupItem.application_id,
      });
      if (applicationToUpdate.groups) {
        applicationToUpdate.groups.unshift(groupItem);
      } else {
        applicationToUpdate.groups = [groupItem];
      }
      this.emitChange();
    });
  }

  deleteGroup(applicationID: string, groupID: string) {
    API.deleteGroup(applicationID, groupID).then(() => {
      const applicationToUpdate = _.findWhere(this.applications as Collection<any>, {
        id: applicationID,
      });
      const newGroups = _.without(
        applicationToUpdate.groups,
        _.findWhere(applicationToUpdate.groups, { id: groupID })
      );

      applicationToUpdate.groups = [...newGroups];
      this.emitChange();
    });
  }

  updateGroup(data: Group) {
    return API.updateGroup(data).then(group => {
      const groupItem = group;
      const applicationToUpdate = _.findWhere(this.applications as Collection<any>, {
        id: groupItem.application_id,
      });
      const index = _.findIndex(applicationToUpdate.groups, { id: groupItem.id });
      applicationToUpdate.groups[index] = groupItem;
      this.emitChange();
    });
  }

  getGroup(applicationID: string, groupID: string) {
    API.getGroup(applicationID, groupID).then(group => {
      const groupItem = group;
      const applicationToUpdate = _.findWhere(this.applications as Collection<any>, {
        id: groupItem.application_id,
      });
      const index = _.findIndex(applicationToUpdate.groups, { id: groupItem.id });

      if (applicationToUpdate) {
        if (applicationToUpdate.groups) {
          if (index >= 0) {
            applicationToUpdate.groups[index] = groupItem;
          } else {
            applicationToUpdate.groups.unshift(groupItem);
          }
        } else {
          applicationToUpdate.groups = [groupItem];
        }
      }
      this.emitChange();
    });
  }

  getGroupVersionCountTimeline(applicationID: string, groupID: string, duration: string) {
    // Not much value here, but we still abstract the API call in case we
    // want to e.g. cache the results in the future.
    return API.getGroupVersionCountTimeline(applicationID, groupID, duration);
  }

  getGroupStatusCountTimeline(applicationID: string, groupID: string, duration: string) {
    // Not much value here, but we still abstract the API call in case we
    // want to e.g. cache the results in the future.
    return API.getGroupStatusCountTimeline(applicationID, groupID, duration);
  }

  // Channels

  createChannel(data: Channel) {
    return API.createChannel(data).then(channel => {
      const channelItem = channel;
      this.getAndUpdateApplication(channelItem.application_id);
    });
  }

  deleteChannel(applicationID: string, channelID: string) {
    API.deleteChannel(applicationID, channelID).then(() => {
      this.getAndUpdateApplication(applicationID);
    });
  }

  updateChannel(data: Channel) {
    return API.updateChannel(data).then(channel => {
      const channelItem = channel;
      this.getAndUpdateApplication(channelItem.application_id);
    });
  }

  // Packages

  createPackage(data: Partial<Package>) {
    return API.createPackage(data).then(packageItem => {
      const newpackage = packageItem;
      this.getAndUpdateApplication(newpackage.application_id);
    });
  }

  deletePackage(applicationID: string, packageID: string) {
    API.deletePackage(applicationID, packageID).then(() => {
      this.getAndUpdateApplication(applicationID);
    });
  }

  updatePackage(data: Partial<Package>) {
    return API.updatePackage(data).then(packageItem => {
      const updatedpackage = packageItem;
      this.getAndUpdateApplication(updatedpackage.application_id);
    });
  }
}

export default ApplicationsStore;
