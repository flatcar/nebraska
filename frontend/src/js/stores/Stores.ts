import ActivityStore from './ActivityStore';
import ApplicationsStore from './ApplicationsStore';
import GroupChartsStore from './GroupChartsStore';

export const applicationsStore = new ApplicationsStore;
export const activityStore = new ActivityStore;
export const groupChartStore = new GroupChartsStore();
