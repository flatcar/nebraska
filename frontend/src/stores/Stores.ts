import { Context, createContext } from 'react';

import ActivityStore from './ActivityStore';
import ApplicationsStore from './ApplicationsStore';
import GroupChartsStore from './GroupChartsStore';

interface Stores {
  applicationsStore: ApplicationsStore;
  activityStore: ActivityStore;
  groupChartStore: GroupChartsStore;

  applicationsStoreContext: Context<ApplicationsStore>;
  activityStoreContext: Context<ActivityStore>;
  groupChartStoreContext: Context<GroupChartsStore>;
}
let stores: Stores | undefined;

export function getStores(noRefresh?: boolean): Stores {
  if (stores === undefined) {
    const applicationsStore = new ApplicationsStore(noRefresh);
    const activityStore = new ActivityStore(noRefresh);
    const groupChartStore = new GroupChartsStore();

    const applicationsStoreContext = createContext(applicationsStore);
    const activityStoreContext = createContext(activityStore);
    const groupChartStoreContext = createContext(groupChartStore);

    stores = {
      applicationsStore,
      activityStore,
      groupChartStore,
      applicationsStoreContext,
      activityStoreContext,
      groupChartStoreContext,
    };
  }
  return stores;
}

export function applicationsStore(noRefresh?: boolean) {
  return getStores(noRefresh).applicationsStore;
}

export function activityStore(noRefresh?: boolean) {
  return getStores(noRefresh).activityStore;
}

export function groupChartStore(noRefresh?: boolean) {
  return getStores(noRefresh).groupChartStore;
}

export function groupChartStoreContext() {
  return getStores().groupChartStoreContext;
}
