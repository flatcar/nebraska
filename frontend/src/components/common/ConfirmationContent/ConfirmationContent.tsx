import { useTranslation } from 'react-i18next';

import { applicationsStore } from '../../../stores/Stores';

function ConfirmationContent(props: {
  data: {
    appID: string;
    groupID?: string;
    channelID?: string;
    packageID?: string;
    type: string;
  };
}) {
  const { t } = useTranslation();

  function processClick() {
    if (props.data.type === 'application') {
      applicationsStore().deleteApplication(props.data.appID);
    } else if (props.data.type === 'group' && props.data.groupID !== undefined) {
      applicationsStore().deleteGroup(props.data.appID, props.data.groupID);
    } else if (props.data.type === 'channel' && props.data.channelID !== undefined) {
      applicationsStore().deleteChannel(props.data.appID, props.data.channelID);
    } else if (props.data.type === 'package' && props.data.packageID !== undefined) {
      applicationsStore().deletePackage(props.data.appID, props.data.packageID);
    }
  }

  return (
    <div className="popover-content" {...props}>
      {t('common|confirmation_prompt')}
      <p className="button-group">
        <button type="button" className="confirm-dialog-btn-abord">
          {t('frequent|no')}
        </button>
        <button type="button" className="confirm-dialog-btn-confirm" onClick={processClick}>
          {t('frequent|yes')}
        </button>
      </p>
    </div>
  );
}

export default ConfirmationContent;
