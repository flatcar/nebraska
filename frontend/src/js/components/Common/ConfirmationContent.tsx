import React from 'react';
import { applicationsStore } from '../../stores/Stores';

function ConfirmationContent(props: {data: {
  appID: string;
  groupID?: string;
  channelID?: string;
  packageID?: string;
  type: string;
};}) {
  function processClick() {
    if (props.data.type === 'application') {
      applicationsStore.deleteApplication(props.data.appID);
    } else if (props.data.type === 'group') {
      applicationsStore.deleteGroup(props.data.appID, props.data.groupID as string);
    } else if (props.data.type === 'channel') {
      applicationsStore.deleteChannel(props.data.appID, props.data.channelID as string);
    } else if (props.data.type === 'package') {
      applicationsStore.deletePackage(props.data.appID, props.data.packageID as string);
    }
  }

  return (
    <div className="popover-content" {...props}>
      Are you sure ... ?
      <p className="button-group">
        <button type="button" className="confirm-dialog-btn-abord">No</button>
        <button type="button" className="confirm-dialog-btn-confirm" onClick={processClick}>Yes</button>
      </p>
    </div>
  );
}

export default ConfirmationContent;
