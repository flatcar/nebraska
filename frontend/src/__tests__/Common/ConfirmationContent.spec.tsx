import '../../i18n/config.ts';

import { act, fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import ConfirmationContent from '../../components/common/ConfirmationContent';
import { applicationsStore } from '../../stores/Stores';

const mockResolver = vi.fn();
describe('Confirmation Content', () => {
  const minProps = {
    data: {
      type: 'application',
      appID: '123',
    },
  };
  it('should render Confirmation Content correctly', () => {
    const { asFragment } = render(<ConfirmationContent data={minProps.data} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should call delete handler function on yes confirmation', async () => {
    applicationsStore().deleteApplication = mockResolver;
    const deleteApp = applicationsStore().deleteApplication;
    const { getByText } = render(<ConfirmationContent data={minProps.data} />);
    await act(async () => fireEvent.click(getByText('Yes')));
    expect(mockResolver).toHaveBeenCalled();
    applicationsStore().deleteApplication = deleteApp;
  });
});
