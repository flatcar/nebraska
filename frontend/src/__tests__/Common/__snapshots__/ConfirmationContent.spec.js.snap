// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirmation Content should render Confirmation Content correctly 1`] = `
<DocumentFragment>
  <div
    channel="[object Object]"
    class="popover-content"
    data="[object Object]"
  >
    Are you sureâ€¦ ?
    <p
      class="button-group"
    >
      <button
        class="confirm-dialog-btn-abord"
        type="button"
      >
        No
      </button>
      <button
        class="confirm-dialog-btn-confirm"
        type="button"
      >
        Yes
      </button>
    </p>
  </div>
</DocumentFragment>
`;
